src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/MovementService.kt

```diff
@@ -1,12 +1,59 @@
 package uk.gov.justice.digital.hmpps.welcometoprison.model
 
+import org.slf4j.LoggerFactory
 import org.springframework.stereotype.Service
+import uk.gov.justice.digital.hmpps.welcometoprison.model.MoveType.PRISON_TRANSFER
 import uk.gov.justice.digital.hmpps.welcometoprison.model.basm.BasmService
 import uk.gov.justice.digital.hmpps.welcometoprison.model.prison.PrisonService
+import uk.gov.justice.digital.hmpps.welcometoprison.model.prisonersearch.PrisonerSearchService
+import uk.gov.justice.digital.hmpps.welcometoprison.model.prisonersearch.response.MatchPrisonerResponse
 import java.time.LocalDate
 
 @Service
-class MovementService(val basmService: BasmService, val prisonService: PrisonService) {
+class MovementService(
+  val basmService: BasmService,
+  val prisonService: PrisonService,
+  val prisonerSearchService: PrisonerSearchService
+) {
   fun getMovements(agencyId: String, date: LocalDate) =
     basmService.getMoves(agencyId, date, date) + prisonService.getMoves(agencyId, date)
+
+  fun getMovementsMatchedWithPrisoner(agencyId: String, date: LocalDate): List<Movement> {
+    val movements = getMovements(agencyId, date)
+    return getMovementsMatchedWithPrisoner(movements)
+  }
+
+  private fun getMovementsMatchedWithPrisoner(movements: List<Movement>): List<Movement> {
+    fun findPrisonerMatch(identifier: String?, identifierName: String): MatchPrisonerResponse? {
+      val matchesByPrisonNumber = identifier?.let { prisonerSearchService.matchPrisoner(it) } ?: emptyList()
+      if (matchesByPrisonNumber.size > 1) log.warn("Multiple matched Prison records for a Movement by $identifierName. There are ${matchesByPrisonNumber.size} matched Prison records for $identifier")
+      return matchesByPrisonNumber.firstOrNull()
+    }
+    fun decorateMovementWithPrisonerMatch(movement: Movement): Movement {
+      var prisonNumberToUse = movement.prisonNumber
+      var pncToUse = movement.pncNumber
+
+      val matchByPrisonNumber = findPrisonerMatch(movement.prisonNumber, "Prison Number")
+      val matchByPncNumber = findPrisonerMatch(movement.pncNumber, "PNC Number")
+
+      if (matchByPrisonNumber === null && matchByPncNumber === null) prisonNumberToUse = null
+      if (matchByPrisonNumber === null && matchByPncNumber !== null) prisonNumberToUse = matchByPncNumber.prisonerNumber
```

Comment:
For these 3 ifs, you will need to check that the values in the results match rather than just the presence of a result.

If you get results from both prison number and pnc searches AND the results have the same prison numbers and PNCs then safe to use both these ids

If you get results from both pnc and prison number search but they don't match - set both values to null

If you get results from one type of search but not the other then use prison number and pnc from the single found result

If you get no results back then should be safe to use the PNC number but clear out the prison number - the prison number must be incorrect but the PNC number could be for a new prisoner

Reply:
Added in an extra two ifs here, think we still need the current ones to cover the case when we only get one match. Suppose we need a decision on whether we still trust Nomis more when we only get one match? I think we do